govcd/api_vcd_versions.go

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"sort"
+
+	semver "github.com/hashicorp/go-version"
+
+	"github.com/vmware/go-vcloud-director/v2/util"
+)
+
+type VersionInfo struct {
+	Version    string `xml:"Version"`
+	LoginUrl   string `xml:"LoginUrl"`
+	Deprecated bool   `xml:"deprecated,attr,omitempty"`
+}
+
+type VersionInfos []VersionInfo
+
+type SupportedVersions struct {
+	VersionInfos `xml:"VersionInfo"`
+}
+
+// APIMaxVerIs allows to compare against maximum vCD supported API version. It will always
+// return false until the client has not done Authenticate(). Can be useful to validate
+// what vCD version is actually running.
+//
+// Format: ">= 27.0, < 32.0", ">= 30.0", "= 27.0"
+//
+// vCD version mapping to API version support https://code.vmware.com/doc/preview?id=8072
+func (vdcCli *VCDClient) APIMaxVerIs(versionConstraint string) bool {
+	util.Logger.Printf("[TRACE] checking max API version against constraints '%s'", versionConstraint)
+	maxVersion, err := vdcCli.maxSupportedVersion()
+	if err != nil {
+		util.Logger.Printf("[ERROR] unable to find max supported version : %s", err)
+		return false
+	}
+
+	isSupported, err := vdcCli.apiVerMatchesConstraint(maxVersion, versionConstraint)
+	if err != nil {
+		util.Logger.Printf("[ERROR] unable to find max supported version : %s", err)
+		return false
+	}
+
+	return isSupported
+}
+
+// APICurVerIs allows to compare against currently used API version. Can be useful to validate
+// if a certain feature can be used or not.
+//
+// Format: ">= 27.0, < 32.0", ">= 30.0", "= 27.0"
+//
+// vCD version mapping to API version support https://code.vmware.com/doc/preview?id=8072
+func (vdcCli *VCDClient) APICurVerIs(versionConstraint string) bool {
```

Comment:
`APIMaxVerIs` is clear, but I find purpose of `APICurVerIs` confusing. A deeper description may be helpful for someone who is seeing these functions first time.

> // APICurVerIs allows to compare against currently used API version
Used by what? Isn't `APIMaxVerIs` the currently used one?

Reply:
I see this is not clear. Trying to rephrase.

* `APICurVerIs` compares against  currently used API version `VCDClient.Client.APIVersion` variable (the API versions which client is set to use).

* `APIMaxVerIs` compares against maximum supported version from `/api/versions` (not necessarily the currently used one). This allows to check what is the maximum API version that vCD instance supports and can be used to guess vCD product version. Like API `31.0` support was first introduced in vCD 9.5 (as per https://code.vmware.com/doc/preview?id=8072). Therefore`if APIMaxVerIs(">= 31.0")` implies that you have vCD 9.5 or newer running inside.

Reply:
Your suggested descriptions are indeed clearer! Now, I'm a proponent of making things clear from function names where possible, so here's a suggestion growing on yours. What about renaming the function as follows:
`APICurVerIs` -> `APICliVerIs `

Reply:
Even better `APIClientVersionIs`, as `Cli` may not always convey the right idea. Let's make it clear that it's the client version and not anything else, such as a `CLI` (command line interface).

Reply:
I like this suggestion (and the commit that just came with it), but then let's make the other function name consistent in style as well:

- `APIMaxVerIs` -> `APIVCDVersionIs` ?
- `APICurVerIs` -> `APIClientVersionIs` (done)

Reply:
Still I'd like some word like `Max` or similar to mention that we're comparing against maximum/biggest version as vCD returns more than supported version. `APIVCDMaxVersionIs`?

Reply:
Fantastic! It turns out to be even of same length :))

`APIVCDMaxVersionIs`
`APIClientVersionIs`

Reply:
I agree due one more reason - using full words ant not shortened ones `Ver`

Reply:
If APIClientVersionIs is check against hardcoded value in of our client? why ist's fetching versions? - or I got something wrong?

Reply:
> If APIClientVersionIs is check against hardcoded value in of our client? why ist's fetching versions? - or I got something wrong?

Very good catch. Removed that.